### PR TITLE
Add clippy to CI

### LINKED ISSUE #25: Add Rust Clippy to CI
## 📚 Context

[Clippy](https://github.com/rust-lang/rust-clippy) is a linter for Rust. We can use it to improve our code quality.

## ✔️ Solution

First, we need to define a sensible configuration for Clippy. We probably can ignore some warnings that don't apply to our code. On the other hand, all warnings that we want to enforce should be considered errors.

Then, we need to fix the existing errors so we can create a new CI workflow that runs Clippy every time. We should evaluate whether it is best to create a new workflow or add Clippy to an existing one.

## 📈 Subtasks

- [x] Configure Clippy
- [x] Fix errors
- [x] Add it to CI

### PR DESCRIPTION
**Pull request by marcelstanley**
#### TL;DR:

**The changes made here are up for discussion**, including the commit messages themselves.

Most of the warnings reported by Clippy were fixed. Besides that, there are some warnings that were suppressed (see `chore` commits) via configuration, mostly because fixing them might take more time.

---

The `ubuntu-latest` image used by CI relies on an outdated version of `clippy`, which makes [the linter fail](https://github.com/cartesi/rollups/actions/runs/5556711553/jobs/10149567440) due to a convoluted [*clone_double_ref* error](https://users.rust-lang.org/t/clone-double-ref/92612), which has been dropped in the most recent `clippy` version [1].

By simply updating `rustup` [1], [clippy passes](https://github.com/cartesi/rollups/actions/runs/5556801966/jobs/10149774799), even though there are several warnings, which are mostly addressed by the commits in this PR.

In [the latest CI run](https://github.com/cartesi/rollups/actions/runs/5557061014/jobs/10150367811?pr=159), with all changes made, there are still the following outstanding warnings:

- [*new_ret_no_self*](https://rust-lang.github.io/rust-clippy/master/index.html#/new_ret_no_self) - either not applicable or costly to fix due to the current business logic;
- [*module_inception*](https://rust-lang.github.io/rust-clippy/master/index.html#/module_inception) - affects code generated via `state-fold-types` and could not be suppressed because the suggested configuration variable did not work under such condition.

---

[1] Expected `rust`-related versions:
```shell
$ rustup -V && rustc -V && cargo -V && cargo clippy -V
rustup 1.26.0 (5af9b9484 2023-04-05)
info: This is the version for the rustup toolchain manager, not the rustc compiler.
info: The currently active `rustc` version is `rustc 1.71.0 (8ede3aae2 2023-07-12)`
rustc 1.71.0 (8ede3aae2 2023-07-12)
cargo 1.71.0 (cfd3bbd8f 2023-06-08)
clippy 0.1.71 (8ede3aa 2023-07-12)
``` 

---

Closes cartesi/rollups#25

### CODE REVIEW

**Review by torives — COMMENTED** [review]
Regarding the remaining warnings, I propose we create an issue for each and postpone promoting warnings to errors on the CI until we fix them.

The `state-fold` code will soon share the same repo as the other services, so the [module_inception](https://rust-lang.github.io/rust-clippy/master/index.html#/module_inception) issue can be tackled then. And we can discuss how to fix [new_ret_no_self](https://rust-lang.github.io/rust-clippy/master/index.html#/new_ret_no_self) in its own issue.

We'll squash the commits fixing the lint errors/warnings into one, right?

File: .github/workflows/code-quality.yml
[the remark this answers is not in the document]
Comment (marcelstanley) [reply .github/workflows/code-quality.yml:1]: > Regarding the remaining warnings, I propose we create an issue for each and postpone promoting warnings to errors on the CI until we fix them.
> 
Nice, that's the idea!

> The `state-fold` code will soon share the same repo as the other services, so the [module_inception](https://rust-lang.github.io/rust-clippy/master/index.html#/module_inception) issue can be tackled then. And we can discuss how to fix [new_ret_no_self](https://rust-lang.github.io/rust-clippy/master/index.html#/new_ret_no_self) in its own issue.
> 
Great!

> We'll squash the commits fixing the lint errors/warnings into one, right?
Yes, we may. And this will help with keeping consistency for the conventional commits initiative, even though I've preferred splitting them to guide the review process.

> What do you think of also moving the `rustfmt` step to this workflow?
I agree!

**Comment by marcelstanley** [comment]
Rebases changes on top of `next`

**Comment by marcelstanley** [comment]
@cartesi/node-unit, lints [*new_ret_no_self*](https://rust-lang.github.io/rust-clippy/master/index.html#/new_ret_no_self) and [*module_inception*](https://rust-lang.github.io/rust-clippy/master/index.html#/module_inception) were disabled at 68447d7571eb086939c4cc065cc7aa0d40893a80, as neither of them could be supressed via configuration. We should work on fixing the related warnings later on at #202.

Besides that, I'd like to highlight once again that a few other lints have been supressed (see commits below) for various reasons:

- 98370b1929191827a25b37f7d9cbfd52ac646161
- 4d518da7793619fda86fadcf626921419b53ba6b
- 2fdd9161461cfecf8b50ef4a9da91fe0d88e4b8e
- 993d7e2edd5443cc29ddc2be4c1f5b648e7818df

Notice, though, that all of them could have been addressed with somewaht more complex changes in the code and could be part of the forthcoming task #202 as well.

File: .github/workflows/code-quality.yml
[the remark this answers is not in the document]
Comment (marcelstanley) [reply .github/workflows/code-quality.yml:1]: Moved code format task at [92de1c7](https://github.com/cartesi/rollups/pull/159/commits/92de1c78f6e4caea99ac7bd9e49143d7b316fe9d)

[the remark this answers is not in the document]
Comment (torives) [reply .github/workflows/code-quality.yml:1]: The split commits made the review a lot easier btw, thx <3

**Comment by torives** [comment]
https://github.com/cartesi/rollups/commit/624cb110e5ab5f6b3a96b852b496f4d7ac7f00eb: I don't think there's much we can do as this is auto-generated from an external dependency. I would leave it as it is, maybe adding a comment to the specific line in the `clippy.toml` file, so we know where that came from.
https://github.com/cartesi/rollups/commit/d00f2f13676e6f482e84cce4b0a35acaa9a52e67: Maybe I'm just used to it, but I have no issues with "Error" as a suffix of error types.
https://github.com/cartesi/rollups/commit/93ac098f3760d636228c70bfcf6d6c4b8aec6486: This is a good candidate for a refactor and IMO should be included in #202
https://github.com/cartesi/rollups/commit/02f79f93eb9c4907793bd49a5ec8aa6849d3097e: No strong opinions on this one, but I'd follow the linter, just because.

**Review by torives — APPROVED** [review]

**Review by gligneul — APPROVED** [review]
Looks good. Do you plan to squash the commits?

**Comment by marcelstanley** [comment]
> Looks good. Do you plan to squash the commits?

Yes, I will squash all fixes into one.

**Comment by marcelstanley** [comment]
> [624cb11](https://github.com/cartesi/rollups/commit/624cb110e5ab5f6b3a96b852b496f4d7ac7f00eb): I don't think there's much we can do as this is auto-generated from an external dependency. I would leave it as it is, maybe adding a comment to the specific line in the `clippy.toml` file, so we know where that came from. 

Done at 5226b24e

**Comment by marcelstanley** [comment]
> [624cb11](https://github.com/cartesi/rollups/commit/624cb110e5ab5f6b3a96b852b496f4d7ac7f00eb): I don't think there's much we can do as this is auto-generated from an external dependency. I would leave it as it is, maybe adding a comment to the specific line in the `clippy.toml` file, so we know where that came from. 

Done at 5226b24e, @torives.
